src/components/organisms/SidebarContainer.tsx

```diff
@@ -0,0 +1,646 @@
+              openCreateStorageDialog()
+            },
+          },
+        ],
+      })
+    },
+    [
+      workspace,
+      storageMap,
+      t,
+      openWorkspaceEditForm,
+      removeWorkspace,
+      togglePreferencesModal,
+      navigate,
+      openCreateStorageDialog,
+    ]
+  )
+
+  const createNoteByRoute = useCallback(async () => {
+    if (workspace == null) {
+      return
+    }
+    const workspaceId = workspace.id
+    let folderPathname = '/'
+    let tags: string[] = []
+    let baseHrefAfterCreate = `/app/storages/${workspaceId}/notes`
+    switch (routeParams.name) {
+      case 'workspaces.labels.show':
+        tags = [routeParams.tagName]
+        baseHrefAfterCreate = `/app/storages/${workspaceId}/tags/${routeParams.tagName}`
+        break
+      case 'workspaces.notes':
+        if (routeParams.folderPathname !== '/') {
+          folderPathname = routeParams.folderPathname
+          baseHrefAfterCreate = `/app/storages/${workspaceId}/notes${folderPathname}`
+        }
+        break
+    }
+
+    const note = await createNote(workspaceId, {
+      folderPathname,
+      tags,
+    })
+    if (note == null) {
+      return
+    }
+
+    push(`${baseHrefAfterCreate}/${note._id}#new`)
+  }, [workspace, routeParams, push, createNote])
+
+  useEffect(() => {
+    if (hash === '#new') {
+      push({ hash: '' })
+      setImmediate(() => {
+        noteDetailFocusTitleInputEventEmitter.dispatch()
+      })
+    }
+  }, [push, hash])
+
+  useEffect(() => {
+    const handler = () => {
+      createNoteByRoute()
+    }
+    addIpcListener('new-note', handler)
+    return () => {
+      removeIpcListener('new-note', handler)
+    }
+  }, [createNoteByRoute])
+
+  const { toggleShowSearchModal } = useSearchModal()
+
+  useEffect(() => {
+    const handler = () => {
+      toggleShowSearchModal()
+    }
+    addIpcListener('search', handler)
+    return () => {
+      removeIpcListener('search', handler)
+    }
+  }, [toggleShowSearchModal])
+
+  useEffect(() => {
+    setGeneralStatus({ lastSidebarState: sidebarState })
+  }, [sidebarState, setSidebarState, setGeneralStatus])
+
+  const openState = useCallback((state: SidebarState) => {
+    setSidebarState((prev) => (prev === state ? undefined : state))
+  }, [])
+
+  const toolbarRows: SidebarToolbarRow[] = useMemo(() => {
+    if (workspace != null) {
+      return mapToolbarRows(
+        showSpaces,
+        setShowSpaces,
+        openState,
+        openTab,
+        toggleShowingCloudIntroModal,
+        sidebarState,
+        workspace
+      )
+    } else {
+      return [
+        {
+          tooltip: 'Spaces',
+          active: showSpaces,
+          icon: mdiMenu,
+          onClick: () => setShowSpaces((prev) => !prev),
+        },
+      ] as SidebarToolbarRow[]
+    }
+  }, [
+    openState,
+    openTab,
+    showSpaces,
+    sidebarState,
+    workspace,
+    toggleShowingCloudIntroModal,
+  ])
+
+  const localSpaces = values(storageMap)
```

Comment:
Don't use any functions returning a new array or an object. This makes other hooks useless.